### ISSUE TITLE
DAMON monitoring results accuracy improvement ideas discussion (was "Does DAMON monitor unused memory area in paddr mode?")

### ISSUE THREAD

**Issue by honggyukim** [damonitor/damo#13]
Hi SeongJae,

I'm just wondering if DAMON monitors unused memory area in paddr mode. For example, if the system has a single NUMA node with 512GB of DRAM and if a user sets the start and end addr range for the entire physical address of 512GB even if most of its space remains unused.

In this case, does DAMON started to evenly split the 512GB into the minimum nr_region then it starts merge/split based on access patterns but the unused area will be detected as cold together so the most of unused area would be merged into a single cold region?

Could you please confirm if my mental model is correct?

Thanks.

**Comment by sjp38** [comment]
Your understanding is correct.

**Comment by honggyukim** [comment]
Thanks for the confirmation. Since the default `nr_regions/min` is `10` and hmsdk uses `100`, it will make giant regions when the system memory has huge capacity. Then it might be difficult to capture small but very hot areas when it previously belongs to the giant cold region.

I think we may have to consider increasing the `nr_regions/{min,max}` but it will also increase the CPU usage. Considering this, I think we can keep `min` regions low but increase it sometimes to make changes the region formation when the CPU usage is low system-widely.

**Comment by sjp38** [comment]
You're right.  DAMON's adaptive regions adjustment is just a best effort, and has many rooms for improvements.  I believe your theory makes sense.  We indeed planning some optimizations for that.

It has not prioritized so far, though, since we found no real evidence of the problem on realistic workloads.  Please report such findings if you have.

**Comment by sjp38** [comment]
Sorry for late response.

> spawning multiple kdamonds when the target address range is too wide to monitor by a single kdamond.

I believe that's a reasonable use of DAMON.  If the single thread performance of DAMON is not enough for your purpose, and you could invest multiple CPUs for your purpose, yes, you can divide-and-rule the monitoring targets with multiple kdamonds :)

**Comment by sjp38** [comment]
In my humble opinion, "too wide to monitor" on the above [comment](https://github.com/damonitor/damo/issues/13#issuecomment-2427990703) is not incorrect, but might allow wrong interpretation.  Let me clarify my humble opinions little bit more.

The size of reasonably important memory regions would be proportional to size of total memory.  On small systems, mistakenly reclaiming hot 4 KiB region could be a problem.  But on large systems, I don't think that's really an issue.  Only once the system is making wrong decision on a scale, it would be a problem.  So, not the absolute size but relative proportion matters.

DAMON's adaptive regions adjustment mechanism is designed with this assumption.  For example, let's suppose that you were able to find 40 MiB working set on 4 GiB memory system.  If you keep everything same and increase the system memory size to 40 GiB, DAMON might not be able to find the 40 MiB working set.  But if the workload is also scaled following the system memory size increase to have 400 MiB working set, DAMON would find it.  Real result would depend on more factors including sampling/aggregation intervals and memory bandwidth.

Of course, I understand the reasonable proportion ratio would depend on use case, and multiple kdamonds can be used if it is applicable for given use case.

So, if I can rephrase the comment in my style, I'd say "we may have to consider spawning multiple kdamonds when a single kdamond cannot meet the target accuracy level of given use case".

**Comment by sjp38** [comment]
FYI, I shared two simple ideas for better DAMON accuracy: https://lore.kernel.org/20241026215311.148363-1-sj@kernel.org

**Comment by sjp38** [comment]
I also shared a DAMON tuning guide for hot pages detection: https://damonitor.github.io/posts/damon_tuning_guide_for_hot_pages/

**Comment by sjp38** [comment]
I think the original question of this issue is answered, and we're now discussing about ways to improve DAMON's monitoring results accuracy.  What do you think about closing this issue and creating another one?

**Comment by honggyukim** [comment]
Hi SeongJae,

Sorry for the late response and thanks very much for your feedback in details as well.

> In my humble opinion, "too wide to monitor" on the above [comment](https://github.com/damonitor/damo/issues/13#issuecomment-2427990703) is not incorrect, but might allow wrong interpretation. Let me clarify my humble opinions little bit more.

I would like to correct the term "too wide to monitor" to "too many regions to monitor" by a single kdamond because DAMON checks only a single page in a region regardless of the size of the region.

> The size of reasonably important memory regions would be proportional to size of total memory. On small systems, mistakenly reclaiming hot 4 KiB region could be a problem. But on large systems, I don't think that's really an issue. Only once the system is making wrong decision on a scale, it would be a problem. So, not the absolute size but relative proportion matters.

That looks correct but I think differently. The 4 KiB of hot region can be negligible in terms of the scale, but let's think this small region is the only place where entire cores access together for a long time, then the access latency of this region is really important.

> DAMON's adaptive regions adjustment mechanism is designed with this assumption. For example, let's suppose that you were able to find 40 MiB working set on 4 GiB memory system. If you keep everything same and increase the system memory size to 40 GiB, DAMON might not be able to find the 40 MiB working set. But if the workload is also scaled following the system memory size increase to have 400 MiB working set, DAMON would find it. Real result would depend on more factors including sampling/aggregation intervals and memory bandwidth.

I think this can be answered by the same answer just above.

> Of course, I understand the reasonable proportion ratio would depend on use case, and multiple kdamonds can be used if it is applicable for given use case.
> 
> So, if I can rephrase the comment in my style, I'd say "we may have to consider spawning multiple kdamonds when a single kdamond cannot meet the target accuracy level of given use case".

That sounds better, but I would like to slightly rephrase it to "we may have to consider spawning multiple kdamonds when a single kdamond have to scan too many regions when the number of regions have to be increased to increase enough accuracy."

**Comment by honggyukim** [comment]
> I also shared a DAMON tuning guide for hot pages detection: https://damonitor.github.io/posts/damon_tuning_guide_for_hot_pages/

Thanks very much. I'm now focusing on feedback driven auto-tuning instead of this so I will take a look when I can find some time.

> I think the original question of this issue is answered, and we're now discussing about ways to improve DAMON's monitoring results accuracy.

Right.

> What do you think about closing this issue and creating another one? I also just found GitHub provides 'Discussions' feature ([h](https://github.com/damonitor/damo/discussions/)

I'm fine with that.

**Comment by sjp38** [comment]
@honggyukim , I agree to all above your comments.  I don't think we're saying completely different things, but just focusing on slightly different angles.  As I also mentioned on the above comment, all depends on the use case.  Using multiple kdamonds is one of expected usages, and we will continue improving DAMON to better support users various use cases.

> The 4 KiB of hot region can be negligible in terms of the scale, but let's think this small region is the only place where entire cores access together for a long time, then the access latency of this region is really important.

So, this will again depends on use case, and I believe such use case can exist.  Nonetheless, I still cannot think this as general and common case for now.  Let me share more details of my humble expectation for such cases in general.  Please bear in mind with me and enlighten me what I'm missing.

On such large systems in general, CPU cache size would also be large enough to get accommodate such special small hot data.  That is, most of the accesses to the 4 KiB region will cache-hit.  DAMON may mistakenly make it evicted from the cache.  But it would be relatively rare event (once per seconds or minutes), compared to its cache hits (say, millions per second?).  Yes, it could increase 99.9999 percentile latency, but is that really a problem?  In general, I further assume the hot data will be periodically evicted from the cache due to other relatively warm or cold working sets, so DAMON adding one more such rare mistake is no big problem.

Again, the above is my humble expectation of "general" large systems.  Everything depends on user's use case.  But, if you find anything unclear or wrong on the above assumption, please let me know.  It could help DAMON prioritizes TODO works, so that it can evolve in right way in right speed.

**Comment by sjp38** [comment]
The feedback-driven monitoring intervals auto-tuning feature for DAMON is now [available](https://lore.kernel.org/20250303221726.484227-1-sj@kernel.org).  It is expected to be merged into v6.15-rc1.  Support of the feautre from `damo` is also available since v2.6.7 via `--monitoring_intervals_goal` option.

**Comment by sjp38** [comment]
Updating the subject to better describe what we're discussing here.

**Comment by sjp38** [comment]
@honggyukim I think the monitoring intervals auto-tuning could solve many of your concerns.  Could you share what you think about the design, and test results of the feature if you could?

**Comment by honggyukim** [comment]
@sjp38 Sorry, I was busy for another work and also focused on weighted interleave feature at the moment. Maybe I can get back to DAMON from the next month. Thanks for your patience.

**Comment by sjp38** [comment]
Hi @honggyukim , could you please give update on this issue?  Especially, I'm looking for your answer to the question on https://github.com/damonitor/damo/issues/13#issuecomment-2726950879

**Comment by sjp38** [comment]
@honggyukim May I gently ask your response again?

**Comment by sjp38** [comment]
Closing this issue for now, since apparently there is no update.  Please open a new one or comment here to ask re-opening of this if needed.
